rep-0151.rst

```diff
@@ -189,12 +189,108 @@ To support the proposal `B)` less effort is necessary compared to proposal
+     explicitly a `python2` or `python3` shebang, as desired.
+   * From-source builds are using Python 2 by default; users may choose
+     instead Python 3 or to use the dual-build behaviour.
+
+Requirements
+''''''''''''
+
+To support the proposal `C)` a few capabilities must be developed in catkin:
+
+ * It will need to be updated with bilingual awareness around the
+   `catkin_install_python` and `catkin_python_setup` functions.
+   
+ * The existing `PYTHON_EXECUTABLE`, `PYTHON_LIBRARY`, and `PYTHON_INCLUDE_DIR`
+   variables which may be used today to select the preferred Python will need to
+   be supplemented by a separate `PYTHON_ALT_*` vars which control which Python is
+   used for the second `setup.py` invocation. If the `PYTHON_ALT_*` vars are unset,
+   catkin will build only for the default Python (this would be default behaviour
+   in a source workspace, in order to not break Arch and other from-source systems
+   where `/usr/bin/python` is already Python 3).
+   
+ * Similarly, Python nosetests will run twice if the `PYTHON_ALT_*` vars are set.
+   Perhaps a CMake function could be used to disable this behaviour, especially
+   for packages which migrate to Python 3 and don't want to receive test failures for
+   Python 2 issues. For example, `catkin_python_policy(ONLY_3)` or similar.
+ 
+ * In order to do the right thing with conventional `usr/bin/env python3` shebangs,
+   ROS Melodic will need to ship a small wrapper shim which mutates the `PYTHONPATH`
+   variable. This shim would live at `/opt/ros/melodic/bin/python3` and look like
+    ```
+    #!/bin/sh
+    PYTHONPATH=$(echo $PYTHONPATH | sed "s/lib\/python2\.7/lib\/python3/")
+    /usr/bin/python3 $*
+    ```
+   Similarly, ROS N would have to do the reverse operation in a `python2` shim. No
+   shim would be required for ROS O and beyond. This shim could be part of the
+   `catkin` package itself, or potentially live separately.
+
+And in other tools:
+
+ * `rosdep`: choose between Python 2 and Python 3 dependencies.
+ 
+ * `bloom`: For the transitional releases (N and M), the default bloom behaviour will
+   be to depend on both the Python 2 and Python 3 versions of all Python dependencies.
+
+ * `ros_buildfarm` should set the new `PYTHON_ALT_*` vars according to the release
+   plan given above.
+   
+ * Packages which generate compiled Python extensions will require bespoke CMake
+   logic to ensure that these extensions are correctly built and installed twice.
+   Upon examining the individual cases, it's possible that the `catkin` package may
+   be able to provide some tooling to streamline this.
+
 Conclusion
 ==========
 
 While proposal `A)` provides a smoother transition it requires a significantly
 higher development effort.
-Due to the sparse resource available proposal `B)` is being chosen.
+Proposal `B)` requires the least tooling effort, but will be the roughest on the
+community, offering little in the way of a migration plan.
```

Comment:
Any conclusion about proposal `C)`?

Reply:
I didn't want to presume. For now, I'm simply explaining the new option so that others can consider what the costs and implications are. I'm assuming a later change will lock in the selection.

For Clearpath's purposes, we have a large enough body of Python that migrating it all atomically is almost certainly not realistic, so _we_'re probably doing some flavour of this proposal regardless. But we also don't use bloom or ros_buildfarm any more, and we only minimally depend on rosdep. So if OSRF wants to make this the official path forward (as I'm suggesting), there are aspects of the tooling with which we unlikely to offer much assistance.

That said, lots of other commercial users _do_ use the ros_buildfarm/bloom tooling, so there may be others who would support this effort in exchange for an eased Py3 transition path.